labman/db/process.py

```diff
@@ -1379,19 +1473,60 @@ def create(cls, user, plate, plate_name, kappa_hyper_plus_kit, stub_lot,
-                i7_comp = i7_layout[
-                    i7_well.row - 1][i7_well.column - 1].composition
-
-                lib_well = container_module.Well.create(
-                    lib_plate, instance, volume, well.row, well.column)
-                composition_module.LibraryPrepShotgunComposition.create(
-                    instance, lib_well, volume, well.composition,
-                    i5_comp, i7_comp)
+
+            # walk across all the positions on the plate in interleaved order;
+            # magic number 4 = get all positions (all 4 quarters)
+            position_generator = GDNAPlateCompressionProcess. \
+                get_interleaved_quarters_position_generator(
+                    4, plate.plate_configuration.num_rows,
+                    plate.plate_configuration.num_columns)
+
+            combo_index = 0
+            plate_layout = plate.layout  # time-consuming so do only once
+            for p in position_generator:  # each position on interleaved plate
+                # note OUTPUT indices rather than input indices because the
+                # normalized gdna plate we are working from is the SAME SIZE
+                # as the library prep plate, not 1/4 its size.
+                input_well = \
+                    plate_layout[p.output_row_index][p.output_col_index]
+                # completely empty wells (represented as Nones) are ignored
+                if input_well is not None:
+                    curr_combo = idx_combos[combo_index]
+                    # As database entities, these "wells" represent the
+                    # positions of the current i5 and i7 primers on the primer
+                    # set plate maps (NOT on the chosen actual, physical primer
+                    # working plates for these primer sets).  That is why we
+                    # aren't just getting the compositions that go with these
+                    # "wells"--these aren't real, physical wells and their
+                    # compositions aren't real, physical compositions.
+                    i5_well = curr_combo[0].container
+                    i7_well = curr_combo[1].container
+
+                    # While the positions of the relevant i5 and i7 primers
+                    # in the i5 and i7 plate maps have separate database
+                    # entities from the positions of those primers on the
+                    # actual physical primer working plates being used, the
+                    # primers are at the same positions (of course!) in the
+                    # working plates made from a given primer set plate map
+                    # as they are in the plate map itself--duh :)
+                    # Thus, we use the position of the relevant i5 and i7
+                    # primers in the plate map (gotten above) to find the
+                    # real, physical compositions for the primers at those
+                    # positions on the real, physical primer working plates.
```

Comment:
Why the -1? Perhaps worth adding to the comments? Also, is this an expensive step? Just wondering if better to load all the i5-7 compositions before starting this and just use that info ...

Reply:
I'll add a comment on the -1, thanks!  My inclination is to hold off on performance optimizations at this stage unless we specifically notice they cause problems during test (e.g., getting plate.layout within the loop causes tests to fail with a time-out, so that was explicitly addressed) ... frankly, I suspect we'll get plenty of feedback from the wet-lab once this goes into test about which parts are slow, at which point we'll have to decide if we can address them piece-meal or need to do something more sweeping.

Reply:
Yep: In that case, as in the case of plate.layout mentioned above, getting layout repeatedly in the loop caused test failures and thus required immediate optimization :)